### DISCUSSION TITLE
How does EVP_KEY_eq work between providers(default & pkcs11-provider) - is there an agnostic comparison format?

### DISCUSSION THREAD

**Discussion by space88man** [openssl/openssl#23652] [Q&A] [answered]
I am using s_server where the cert is a default provider object (from a PEM file), and the private key is from a 3rd party provider (in this case a pkcs11-provider private key). In the pkcs#11 token I only have a private key, with no matching public key object or certificate object.

`SSL_CTX_use_PrivateKey` calls `X509_check_private_key` which then reaches `EVP_KEY_eq`.
At this point it fails as the HSM does not have a matching pubkey and I can observe pkcs11-provider performing an inventory of objects looking for a matching public key - but I'm not sure if this a limitation of the pkcs11-provider.

How is this actually supposed to work when the providers are different? Is there a provider-agnostic format to perform logical equality or is this an implementation defect in pkcs11-provider?

This is doable in the  engine world: i.e.,  `s_server -cert APemFile -engine pkcs11 -keyform engine  -key SomeEngineKey` works.

pkcs11-provider has a test for s_server but the token it uses has a matching public key so avoids the issue that I am raising here.

https://github.com/latchset/pkcs11-provider/issues/345

**Comment by mattcaswell** [comment, marked as answer]
It's quite complicated and handles provider based and legacy keys.

It does some simple trivial checks first (e.g. if the keys are the same actual pointer then they are equal, if one of the keys is NULL and the other isn't, then they are different, etc).

I'll ignore legacy keys since that adds another layer of complexity. So, lets assume both keys are provider based and that the trivial checks above don't apply. In that case:

1) If both keys are from the same provider then we ask the provider to decide. The provider must supply a keymgmt "match" function for this to work. If not, keys are never considered equal.
2) If the keys have different keymgmt instances (usually because they're from different providers) then we check then if keymgmt2 has a "match" function then we attempt to export the key from keymgmt1 into keymgmt2. For this to work keymgmt1 must have an "export" function and "keymgmt2" must have an "import" function. Once they are both in the same keymgmt, then do step (1) again.
3) If (2) does not work then try the other way around (i.e. export from keymgmt2 into keymgmt1).

**Comment by t8m** [comment]
So, for your concrete use case, where the private key cannot be exported, the pkcs11 provider would have to be able to import the public key and use some way to match it with the private key it holds. How that would be implemented or if it is at all possible given the data stored in the HSM, that is a different matter. One option to workaround this would be to have an option to not do the X509_check_private_key() call and just depend on the configuration being correct. The check call is something that is not actually necessary for the TLS server to be functional. It is just a misconfiguration prevention.

**Comment by space88man** [comment]
Thank you - great insight.

In my use case it turns out to be a bug in pkcs11-provider: it had enough information from the private key (MODULUS/PUBLIC_EXPONENT)  to make the logical equality comparison with the (imported from PEM) public key but continued to search the HSM for a matching public key.
